Gstreamer pipeline is not starting

Gstreamer pipeline is not starting because in __main__.py   lambda preventing to starting pipeline ```loop.run_in_executor(None, lambda: app.start_pipeline()) ``` 
![image](https://github.com/selkies-project/selkies-gstreamer/assets/141987684/d0c6bb74-7796-44c2-9de9-3df0073ff34b)
after deleting lambda pipeline started 
![image](https://github.com/selkies-project/selkies-gstreamer/assets/141987684/dc8220ae-4f47-44ec-be51-ab2c5a5bdb87)

and there is another issue If we change from 

        while self.bus is not None and self.pipeline is not None:
            while self.bus.have_pending():
                msg = self.bus.pop()
                if not self.bus_call(msg):
                    self.bus = None
                    break
            time.sleep(0.1)
            
 to print("started") we see 
![image](https://github.com/selkies-project/selkies-gstreamer/assets/141987684/bb56bc4b-c152-44a0-af30-017d2c8ba7c2)
and that means time.sleep(0.1) triggering infinite sleep and preventing to start the session
If we change time.sleep(0.1) to "break" everything is ok.

docker run --name selkies -it --rm -p 8080:8080 ghcr.io/selkies-project/selkies-gstreamer/gst-py-example:main-ubuntu20.04
Unable to find image 'ghcr.io/selkies-project/selkies-gstreamer/gst-py-example:main-ubuntu20.04' locally
main-ubuntu20.04: Pulling from selkies-project/selkies-gstreamer/gst-py-example
7007490126ef: Pull complete
4199b68575e5: Pull complete
4f4fb700ef54: Pull complete
12f5a0ad6c3b: Pull complete
ebaaf492626f: Pull complete
6754a939ad6e: Pull complete
1e9b00e16b3f: Pull complete
c38ef1eaab2b: Pull complete
19a4851f11ce: Pull complete
10baa9059c62: Pull complete
3a31d8ace2ae: Pull complete
8fd0d5dfd5c9: Pull complete
a8e772851fc1: Pull complete
27981f4645e1: Pull complete
f44ea577f22d: Pull complete
0cbf1c5fdc23: Pull complete
5a13860b5631: Pull complete
Digest: sha256:f64243313cc84e01b0ec1151c4870803c6b5d534cae4d28a37f622b354e01f53
Status: Downloaded newer image for ghcr.io/selkies-project/selkies-gstreamer/gst-py-example:main-ubuntu20.04
_XSERVTransmkdir: ERROR: euid != 0,directory /tmp/.X11-unix will not be created.
screen 0 shmid 0
X Server is ready

(xfce4-session:23): dbind-WARNING **: 15:32:17.898: AT-SPI: Error retrieving accessibility bus address: org.freedesktop.DBus.Error.ServiceUnknown: The name org.a11y.Bus was not provided by any .service files
_IceTransmkdir: ERROR: euid != 0,directory /tmp/.ICE-unix will not be created.
xfce4-session: No SSH authentication agent found
gpg-agent[39]: WARNING: "--write-env-file" is an obsolete option - it has no effect
gpg-agent[39]: directory '/home/user/.gnupg' created
gpg-agent[39]: directory '/home/user/.gnupg/private-keys-v1.d' created

(xfce4-session:23): xfce4-session-WARNING **: 15:32:17.964: gpg-agent returned no PID in the variables

(xfce4-session:23): xfce4-session-WARNING **: 15:32:17.967: xfsm_manager_load_session: Something wrong with /home/user/.cache/sessions/xfce4-session-eb64b65972cb:0, Does it exist? Permissions issue?
gpg-agent[40]: gpg-agent (GnuPG) 2.2.19 started

(xfwm4:41): dbind-WARNING **: 15:32:18.010: AT-SPI: Error retrieving accessibility bus address: org.freedesktop.DBus.Error.ServiceUnknown: The name org.a11y.Bus was not provided by any .service files

(xfwm4:41): GLib-CRITICAL **: 15:32:18.141: g_str_has_prefix: assertion 'prefix != NULL' failed
0:00:00.364464236    28 0x55d228c39260 WARN              cudaloader gstcudaloader.c:169:gst_cuda_load_library: Could not open library libcuda.so.1, libcuda.so.1: cannot open shared object file: No such file or directory
0:00:00.365524640    28 0x55d228c39260 WARN                 nvcodec plugin.c:94:plugin_init: Failed to load cuda library
0:00:00.688787286    28 0x55d228c39260 WARN             msdkcontext gstmsdkcontext.c:177:gst_msdk_context_use_vaapi: Couldn't find a drm device node to open
0:00:00.829391136    28 0x55d228c39260 WARN                 default ges-meta-container.c:236:_set_value:<GESAsset@0x55d22926eae0> Could not set value on item: format-version
0:00:00.830542059    28 0x55d228c39260 WARN                 default ges-meta-container.c:236:_set_value:<GESAsset@0x55d22926eb60> Could not set value on item: format-version
0:00:00.831105415    28 0x55d228c39260 WARN                 default ges-meta-container.c:236:_set_value:<GESAsset@0x55d22926ebe0> Could not set value on item: format-version
0:00:00.833896065    28 0x55d228c39260 WARN               structure gststructure.c:2334:priv_gst_structure_parse_fields: Failed to find delimiter, r=mimetype
0:00:00.861377545    28 0x55d228c39260 WARN      GST_PLUGIN_LOADING gstplugin.c:534:gst_plugin_register_func: plugin "/opt/gstreamer/lib/x86_64-linux-gnu/gstreamer-1.0/validate/libgstvalidatessim.so" failed to initialise
INFO: gst-python install looks OK
xrandr: Failed to get size of gamma for output screen
INFO:gstwebrtc_app_resize:resizing display to 1280x720
INFO:gstwebrtc_app_resize:adding mode 1280x720 to xrandr screen 'screen'
INFO:gstwebrtc_app_resize:creating new xrandr mode: 1280x720 63.75  1280 1328 1360 1440  720 723 728 741 +hsync -vsync
INFO:gstwebrtc_app_resize:adding xrandr mode '1280x720' to screen 'screen'
INFO:gstwebrtc_app_resize:applying xrandr screen 'screen' mode: 1280x720
True
INFO: gst-python install looks OK

(xfwm4:41): xfwm4-WARNING **: 15:32:19.279: Another compositing manager is running on screen 0
WARNING:root:Namespace(addr='0.0.0.0', app_auto_init='true', app_ready_file='/var/run/appconfig/appready', audio_bitrate='64000', audio_channels='2', basic_auth_password='', basic_auth_user='user', coturn_auth_header_name='x-auth-user', coturn_web_uri='', coturn_web_username='selkies-eb64b65972cb', cursor_size='-1', debug=False, debug_cursors='false', enable_audio='true', enable_basic_auth='false', enable_clipboard='true', enable_cursors='true', enable_https='false', enable_resize='true', encoder='x264enc', framerate='30', https_cert='/etc/ssl/certs/ssl-cert-snakeoil.pem', https_key='/etc/ssl/private/ssl-cert-snakeoil.key', json_config='/tmp/selkies_config.json', metrics_port='8000', port='8080', rtc_config_json='/tmp/rtc.json', turn_host='', turn_password='', turn_port='', turn_protocol='udp', turn_shared_secret='', turn_tls='false', turn_username='', uinput_js_socket='', uinput_mouse_socket='', video_bitrate='2000', web_root='/opt/gst-web')
INFO:main:Waiting for streaming app ready
WARNING:main:error fetching coturn RTC config, using DEFAULT_RTC_CONFIG: [Errno -2] Name or service not known
INFO:main:initial server RTC config: {
  "lifetimeDuration": "86400s",
  "iceServers": [
    {
      "urls": [
        "stun:stun.l.google.com:19302"
      ]
    }
  ],
  "blockStatus": "NOT_BLOCKED",
  "iceTransportPolicy": "all"
}
INFO:signaling:Listening on http://0.0.0.0:8080

(xfsettingsd:56): dbind-WARNING **: 15:32:19.471: AT-SPI: Error retrieving accessibility bus address: org.freedesktop.DBus.Error.ServiceUnknown: The name org.a11y.Bus was not provided by any .service files
INFO:webrtc_input:Resetting keyboard modifiers.

(xfce4-panel:57): dbind-WARNING **: 15:32:19.479: AT-SPI: Error retrieving accessibility bus address: org.freedesktop.DBus.Error.ServiceUnknown: The name org.a11y.Bus was not provided by any .service files

(xfsettingsd:56): libupower-glib-WARNING **: 15:32:19.496: Couldn't connect to proxy: Could not connect: No such file or directory

(xfsettingsd:56): libupower-glib-CRITICAL **: 15:32:19.497: up_client_get_lid_is_closed: assertion 'UP_IS_CLIENT (client)' failed

(xfsettingsd:56): GLib-GObject-WARNING **: 15:32:19.498: invalid (NULL) pointer instance

(xfsettingsd:56): GLib-GObject-CRITICAL **: 15:32:19.498: g_signal_connect_data: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed

(xfsettingsd:56): GLib-CRITICAL **: 15:32:19.519: g_str_has_prefix: assertion 'prefix != NULL' failed

(xfsettingsd:56): GLib-GObject-CRITICAL **: 15:32:19.526: g_value_get_string: assertion 'G_VALUE_HOLDS_STRING (value)' failed

(xfsettingsd:56): GLib-GObject-CRITICAL **: 15:32:19.530: g_value_get_string: assertion 'G_VALUE_HOLDS_STRING (value)' failed
INFO:webrtc_input:starting clipboard monitor
INFO:webrtc_input:Found XFIXES version 4.0
INFO:webrtc_input:starting cursor monitor
INFO:webrtc_input:watching for cursor changes
INFO:webrtc_input:sending clipboard content, length: 34

(Thunar:72): dbind-WARNING **: 15:32:19.932: AT-SPI: Error retrieving accessibility bus address: org.freedesktop.DBus.Error.ServiceUnknown: The name org.a11y.Bus was not provided by any .service files
xfce4-panel-Message: 15:32:20.062: Plugin "statusnotifier-7" was not found and has been removed from the configuration

(xfdesktop:77): dbind-WARNING **: 15:32:20.065: AT-SPI: Error retrieving accessibility bus address: org.freedesktop.DBus.Error.ServiceUnknown: The name org.a11y.Bus was not provided by any .service files
xfce4-panel-Message: 15:32:20.070: Plugin "power-manager-plugin-9" was not found and has been removed from the configuration
xfce4-panel-Message: 15:32:20.072: Plugin "notification-plugin-10" was not found and has been removed from the configuration

(wrapper-2.0:78): dbind-WARNING **: 15:32:20.141: AT-SPI: Error retrieving accessibility bus address: org.freedesktop.DBus.Error.ServiceUnknown: The name org.a11y.Bus was not provided by any .service files

(wrapper-2.0:82): dbind-WARNING **: 15:32:20.174: AT-SPI: Error retrieving accessibility bus address: org.freedesktop.DBus.Error.ServiceUnknown: The name org.a11y.Bus was not provided by any .service files

(wrapper-2.0:79): dbind-WARNING **: 15:32:20.200: AT-SPI: Error retrieving accessibility bus address: org.freedesktop.DBus.Error.ServiceUnknown: The name org.a11y.Bus was not provided by any .service files

(wrapper-2.0:82): libactions-WARNING **: 15:32:20.351: Calling CanShutdown failed GDBus.Error:org.xfce.SessionManager.Error.Failed: Could not connect: No such file or directory

(wrapper-2.0:82): libactions-WARNING **: 15:32:20.355: Calling CanRestart failed GDBus.Error:org.xfce.SessionManager.Error.Failed: Could not connect: No such file or directory

(xfce4-session:23): xfce4-session-WARNING **: 15:32:20.373: failed to run script: Failed to execute child process ?/usr/bin/pm-is-supported? (No such file or directory)

(xfce4-session:23): xfce4-session-WARNING **: 15:32:20.385: failed to run script: Failed to execute child process ?/usr/bin/pm-is-supported? (No such file or directory)

(xfce4-session:23): xfce4-session-WARNING **: 15:32:20.393: failed to run script: Failed to execute child process ?/usr/bin/pm-is-supported? (No such file or directory)

(wrapper-2.0:82): Gtk-WARNING **: 15:32:20.414: Negative content width -1 (allocation 1, extents 1x1) while allocating gadget (node button, owner XfceArrowButton)

(wrapper-2.0:79): libnotify-WARNING **: 15:32:20.461: Failed to connect to proxy

(wrapper-2.0:79): Gtk-WARNING **: 15:32:20.478: Negative content width -3 (allocation 1, extents 2x2) while allocating gadget (node button, owner PulseaudioButton)

** (xfdesktop:77): WARNING **: 15:32:20.509: Thumbnailer failed calling GetFlavors

(xfdesktop:77): GLib-GObject-CRITICAL **: 15:32:20.510: g_object_unref: assertion 'G_IS_OBJECT (object)' failed

(xfsettingsd:99): dbind-WARNING **: 15:32:20.565: AT-SPI: Error retrieving accessibility bus address: org.freedesktop.DBus.Error.ServiceUnknown: The name org.a11y.Bus was not provided by any .service files

(xfce4-session:23): xfce4-session-WARNING **: 15:32:20.579: ICE connection 0x55ccb0546dc0 rejected
Failure: Module initialization failed

** (xfdesktop:77): WARNING **: 15:32:20.633: Failed to get system bus: Could not connect: No such file or directory
INFO:signaling:Connected to ('127.0.0.1', 46758)
INFO:signaling:Registered peer '0' at ('127.0.0.1', 46758) with meta: None
INFO:signaling:'0' command 'SESSION 1'
INFO:signaling:'0' command 'SESSION 1'
INFO:signaling:'0' command 'SESSION 1'
INFO:signaling:'0' command 'SESSION 1'
INFO:signaling:'0' command 'SESSION 1'
INFO:signaling:'0' command 'SESSION 1'
INFO:signaling:'0' command 'SESSION 1'
INFO:signaling:'0' command 'SESSION 1'
INFO:signaling:'0' command 'SESSION 1'
INFO:signaling:'0' command 'SESSION 1'
INFO:signaling:'0' command 'SESSION 1'
INFO:signaling:Connected to ('172.17.0.1', 42730)
INFO:signaling:'0' command 'SESSION 1'
INFO:signaling:Registered peer '1' at ('172.17.0.1', 42730) with meta: {'res': '1552x1172', 'scale': 1}
INFO:signaling:'0' command 'SESSION 1'
INFO:signaling:Session from '0' (('127.0.0.1', 46758)) to '1' (('172.17.0.1', 42730))
INFO:main:starting session with meta: {'res': '1552x1172', 'scale': 1}
xrandr: Failed to get size of gamma for output screen
WARNING:main:resizing display from 1280x720 to 1552x1172
xrandr: Failed to get size of gamma for output screen
INFO:gstwebrtc_app_resize:resizing display to 1552x1172
INFO:gstwebrtc_app_resize:adding mode 1552x1172 to xrandr screen 'screen'
INFO:gstwebrtc_app_resize:creating new xrandr mode: 1552x1172 123.75  1552 1600 1632 1712  1172 1175 1185 1206 +hsync -vsync
INFO:gstwebrtc_app_resize:adding xrandr mode '1552x1172' to screen 'screen'
INFO:gstwebrtc_app_resize:applying xrandr screen 'screen' mode: 1552x1172
INFO:gstwebrtc_app:sending remote resolution of: 1552x1172
INFO:main:Setting DPI to: 96
INFO:main:Setting cursor size to: 16



The same issue happening   
[20:35:36] [signalling] Connecting to server.

[20:35:38] [signalling] Registering with server, peer ID: 1

[20:35:40] [signalling] Registered with server.

[20:35:40] [signalling] Waiting for video stream.

[20:36:1] [webrtc] [ERROR] attempt to send data channel message before channel was open.

[20:37:6] [webrtc] [ERROR] attempt to send data channel message before channel was open.

[20:37:10] [webrtc] [ERROR] attempt to send data channel message before channel was open.

[20:37:14] [webrtc] [ERROR] attempt to send data channel message before channel was open.

--- a/.github/actions/build_and_publish_image/action.yaml
+++ b/.github/actions/build_and_publish_image/action.yaml
@@ -36,6 +36,10 @@ inputs:
     default: "false"
     description: Use cached layers from pre-published image with version 1 (true/false)
     required: false
+  push_image:
+    default: "true"
+    description: Push image to registry
+    required: false
 
 runs:
   using: composite
@@ -65,10 +69,14 @@ runs:
         else
           (cd ${{ inputs.image_source_directory }} && docker build $BUILD_ARGS -f ${{ inputs.dockerfile }} -t $IMAGE_TAG_1 .)
         fi
-        docker push $IMAGE_TAG_1
+        if [[ "${{ inputs.push_image }}" == "true" ]]; then
+          docker push $IMAGE_TAG_1
+        fi
 
         if [ '${{ inputs.image_version_2 }}' != '' ]; then
           export IMAGE_TAG_2="${IMAGE_TAG_URL_PREFIX}${{ inputs.image_name }}:${{ inputs.image_version_2 }}${{ inputs.image_version_suffix }}"
           docker tag $IMAGE_TAG_1 $IMAGE_TAG_2
-          docker push $IMAGE_TAG_2
+          if [[ "${{ inputs.push_image }}" == "true" ]]; then
+            docker push $IMAGE_TAG_2
+          fi
         fi

--- a/.github/workflows/build_changed_images.yaml
+++ b/.github/workflows/build_changed_images.yaml
@@ -2,12 +2,9 @@
 # License, v. 2.0. If a copy of the MPL was not distributed with this
 # file, You can obtain one at https://mozilla.org/MPL/2.0/.
 
-name: Build & publish changed images
+name: Build changed images
 
 on:
-  push:
-    branches:
-      - main
   pull_request:
       branches: main
       paths-ignore:
@@ -85,7 +82,8 @@ jobs:
           github_username: $GITHUB_ACTOR
           image_name: ${{ matrix.name }}
           image_source_directory: ${{ matrix.source_directory }}
-          image_version_1: $GITHUB_REF_NAME${{ matrix.version_suffix }}
+          image_version_1: pr${{ matrix.version_suffix }}
+          push_image: "false"
 
   # Note: When modifying this job, copy modifications to all other workflows' image jobs.
   changed_example_images:
@@ -96,7 +94,8 @@ jobs:
         include:
           - name: gst-py-example
             version_suffix: -ubuntu20.04
-            build_args: PACKAGE_VERSION=0.0.0.dev0;UBUNTU_RELEASE=20.04;GSTREAMER_BASE_IMAGE_RELEASE=${{ github.ref_name }};PY_BUILD_IMAGE=ghcr.io/selkies-project/selkies-gstreamer/py-build:${{ github.ref_name }};WEB_IMAGE=ghcr.io/selkies-project/selkies-gstreamer/gst-web:${{ github.ref_name }}
+            push_image: "false"
+            build_args: PACKAGE_VERSION=0.0.0.dev0;UBUNTU_RELEASE=20.04;GSTREAMER_BASE_IMAGE_RELEASE=pr;PY_BUILD_IMAGE=ghcr.io/selkies-project/selkies-gstreamer/py-build:pr;WEB_IMAGE=ghcr.io/selkies-project/selkies-gstreamer/gst-web:pr
             dockerfile: Dockerfile.example
             source_directory: .
             source_files_for_diff: |
@@ -110,7 +109,8 @@ jobs:
 
           - name: gst-py-example
             version_suffix: -ubuntu22.04
-            build_args: PACKAGE_VERSION=0.0.0.dev0;UBUNTU_RELEASE=22.04;GSTREAMER_BASE_IMAGE_RELEASE=${{ github.ref_name }};PY_BUILD_IMAGE=ghcr.io/selkies-project/selkies-gstreamer/py-build:${{ github.ref_name }};WEB_IMAGE=ghcr.io/selkies-project/selkies-gstreamer/gst-web:${{ github.ref_name }}
+            push_image: "false"
+            build_args: PACKAGE_VERSION=0.0.0.dev0;UBUNTU_RELEASE=22.04;GSTREAMER_BASE_IMAGE_RELEASE=pr;PY_BUILD_IMAGE=ghcr.io/selkies-project/selkies-gstreamer/py-build:pr;WEB_IMAGE=ghcr.io/selkies-project/selkies-gstreamer/gst-web:pr
             dockerfile: Dockerfile.example
             source_directory: .
             source_files_for_diff: |
@@ -144,4 +144,5 @@ jobs:
           github_username: $GITHUB_ACTOR
           image_name: ${{ matrix.name }}
           image_source_directory: ${{ matrix.source_directory }}
-          image_version_1: $GITHUB_REF_NAME${{ matrix.version_suffix }}
+          image_version_1: pr${{ matrix.version_suffix }}
+          push_image: "false"

--- a/src/selkies_gstreamer/__main__.py
+++ b/src/selkies_gstreamer/__main__.py
@@ -569,7 +569,7 @@ def main():
                     on_resize_handler(meta["res"])
                 if meta["scale"]:
                     on_scaling_ratio_handler(meta["scale"])
-        loop.run_in_executor(None, lambda: app.start_pipeline())
+        app.start_pipeline()
     signalling.on_session = on_session_handler
 
     # Initialize the Xinput instance
@@ -779,6 +779,7 @@ def main():
         loop.run_in_executor(None, lambda: system_mon.start())
 
         while True:
+            asyncio.ensure_future(app.handle_bus_calls(), loop=loop)
             loop.run_until_complete(signalling.connect())
             loop.run_until_complete(signalling.start())
             app.stop_pipeline()

--- a/src/selkies_gstreamer/gstwebrtc_app.py
+++ b/src/selkies_gstreamer/gstwebrtc_app.py
@@ -1126,15 +1126,19 @@ class GSTWebRTCApp:
 
         logger.info("pipeline started")
 
+    async def handle_bus_calls(self):
         # Start bus call loop
-        self.bus = self.pipeline.get_bus()
-        while self.bus is not None and self.pipeline is not None:
-            while self.bus.have_pending():
-                msg = self.bus.pop()
-                if not self.bus_call(msg):
-                    self.bus = None
-                    break
-            time.sleep(0.1)
+        running = True
+        bus = None
+        while running:
+            if self.pipeline is not None:
+                bus = self.pipeline.get_bus()
+            if bus is not None:
+                while bus.have_pending():
+                    msg = bus.pop()
+                    if not self.bus_call(msg):
+                        running = False
+            await asyncio.sleep(0.1)
 
     def stop_pipeline(self):
         logger.info("stopping pipeline")

